nautilus_core/common/src/throttler/mod.rs

```diff
@@ -0,0 +1,387 @@
+        assert_eq!(inner.sent_count, 5);
+    }
+
+    #[rstest]
+    fn test_buffering_used_before_limit_when_halfway_returns_half(
+        mut test_throttler_buffered: TestThrottler,
+    ) {
+        let throttler = &mut test_throttler_buffered.throttler;
+
+        for _ in 0..3 {
+            throttler.send("MESSAGE".to_string());
+        }
+
+        let inner = throttler.inner.borrow();
+        assert_eq!(inner.used(), 0.6);
+        assert_eq!(inner.recv_count, 3);
+        assert_eq!(inner.sent_count, 3);
+    }
+
+    #[rstest]
+    fn test_buffering_refresh_when_at_limit_sends_remaining_items(
+        mut test_throttler_buffered: TestThrottler,
+    ) {
+        let throttler = &mut test_throttler_buffered.throttler;
+
+        for _ in 0..6 {
+            throttler.send("MESSAGE".to_string());
+        }
+
+        // Advance time and process events
+        {
+            let mut clock = test_throttler_buffered.clock.borrow_mut();
+            let time_events = clock.advance_time(1_000_000_000.into(), true);
+            for each_event in clock.match_handlers(time_events) {
+                drop(clock); // Release the mutable borrow
+
+                each_event.callback.call(each_event.event);
+
+                // Re-borrow the clock for the next iteration
+                clock = test_throttler_buffered.clock.borrow_mut();
+            }
+        }
+
+        // Assert final state
+        {
+            let inner = throttler.inner.borrow();
+            assert_eq!(inner.used(), 0.0);
+            assert_eq!(inner.recv_count, 6);
+            assert_eq!(inner.sent_count, 6);
+            assert_eq!(inner.qsize(), 0);
+        }
+    }
+
+    #[rstest]
+    fn test_buffering_send_message_after_buffering_message(
+        mut test_throttler_buffered: TestThrottler,
+    ) {
+        let throttler = &mut test_throttler_buffered.throttler;
+
+        for _ in 0..6 {
+            throttler.send("MESSAGE".to_string());
+        }
+
+        // Advance time and process events
+        {
+            let mut clock = test_throttler_buffered.clock.borrow_mut();
+            let time_events = clock.advance_time(1_000_000_000.into(), true);
+            for each_event in clock.match_handlers(time_events) {
+                drop(clock); // Release the mutable borrow
+
+                each_event.callback.call(each_event.event);
+
+                // Re-borrow the clock for the next iteration
+                clock = test_throttler_buffered.clock.borrow_mut();
+            }
+        }
+
+        for _ in 0..6 {
+            throttler.send("MESSAGE".to_string());
+        }
+
+        // Assert final state
+        {
+            let inner = throttler.inner.borrow();
+            // Some Logical Error? It should not be 0.
+            assert_eq!(inner.used(), 0.0);
+            assert_eq!(inner.recv_count, 12);
+            assert_eq!(inner.sent_count, 12);
+            assert_eq!(inner.qsize(), 0);
+        }
+    }
```

Comment:
Based on my understanding, it seems that `inner.used()` @ L290 should not be equal to `0.0`. I suspect there may be a logical error in the implementation. I'm currently working on a fix for this issue.

Reply:
fixed